docs/src/adr/005adr-json.md

```diff
@@ -32,23 +32,93 @@ The following classes are serializable:
 ```
 {
-  kind: "NameEx"
+  "kind": "NameEx"
   ...
 }
 ```
 
-## 3. Serializing typed entities
+## 3. Serializing tagged entities
 
-Serializations of typed entities will have an additional field named `type`, containing the type of the expression (see [ADR-002][], [ADR-004][] for a description of our type system and the syntax for types-as-string-annotations respectively). For example, the integer literal `1` is represented by a `ValEx`, which has type `Int` and is serialized as follows:
+Serializations of entities annotated with a `TypeTag` will have an additional field named `type`, containing the type of the expression (see [ADR-002][], [ADR-004][] for a description of our type system and the syntax for types-as-string-annotations respectively), if the tag is `Typed`, or `Untyped` otherwise. For example, the integer literal `1` is represented by a `ValEx`, which has type `Int` and is serialized as follows:
 ```
 {
-  kind: "NameEx"
-  type: "Int"
+  "kind": "ValEx",
+  "type": "Int"
   ...
 }
 ```
+in the typed encoding, or 
+```
+{
+  "kind": "ValEx",
+  "type": "Untyped"
+  ...
+}
+```
+in the untyped encoding.
 
-## 4. Serialization rules
+## 4. Source information
+Entities in the internal representation are usually annotated with source information, of the form `{filename}:{startLine}:{startColumn}-{endLine}:{endColumn}`, relating them to a file range in the provided specification (from which they may have been transformed as part of preprocessing).
+JSON encodings may, but are not required to, contain a `source` providing this information, of the following shape:
+```
+{
+  "source": {
+    "filename" : <FILENAME>,
+    "from" : {
+      "line" : <STARTLINE>,
+      "column" : <STARTCOLUMN>
+    },
+    "to" : {
+      "line" : <ENDLINE>,
+      "column" : <ENDCOLUMN>
+    }
+  }
+}
+```
+or
+```
+{
+  "source": "UNKNOWN"
+}
+```
+if no source information is available (e.g. for expressions generated purely by Apalache).
+Serializations generated by Apalache are always guaranteed to contain a `source` field entry.
+
+Example:
+```
+{
+  "kind" : "NameEx",
+  "type" : "Int",
+  "name" : "myName",
+  "source": {
+    "filename" : "MyModule.tla",
+    "from" : {
+      "line" : 3,
+      "column" : 5
+    },
+    "to" : {
+      "line" : 3,
+      "column" : 10
+    }
+  }
+}
+```
+
+## 5. Version information
+JSON serializations of `TlaModule`s contain a filed named `version`, holding a string representation of the current JSON encoding version, shaped `{major}.{minor}`.
+This document defines JSON Version 1.0. If and when a different JSON version is defined, this document will be updated accordingly.
+Apalache may refuse to import, or trigger warnings for, JSON objects with obsolete versions of the encoding in the future.
+Example:
+```
+{
+  "kind": "TlaModule",
+  "name": "MyModule",
+  "version": "1.0"
```

Comment:
I think we should introduce a different root object for the json format. We may want to add meta information to it later. So I would suggest adding another object on top of `TlaModule`, e.g.:

```tla
{
  "name": "ApalacheIR",
  "version": "1.0",
  "modules": [{
    "kind": "TlaModule",
    "name": "MyModule",
    ...
  }]
}
```